### DISCUSSION TITLE
Wiele funkcji, jak najlepiej ogarnąć lokalny development

### DISCUSSION THREAD

**Discussion by GKotfis** [AzureCommunityPL/group#340] [Pomoc]
Próbuję polepszyć trochę Developer Experience w zakresie pracy z funkcjami.
W aktualnym projekcie mamy 5 funkcji (osobne projekty, każda na własnym porcie), które zależą od siebie. Często, aby uruchomić cały kombajn, trzeba postawić je wszystkie.
Jak najlepiej to ogarnąć? Zależy mi, aby metoda nie była powiązana z żadnym IDE i systemem operacyjnym.

1. Skrypt, który woła `func start` dla każdej funkcji. Jak wtedy ogarnąć logi na konsoli? 5 okienek na raz będzie upierdliwe
2. Jeśli powiązać to z IDE to można skorzystać w VisualStudio z `multiproject` a w Rider z `Compound run`. Zostaje jeszcze VSCode. Różni programiści są w zespole i każdy ma swój toolset.
3. Może wpakować je do `Dockera` i odpalić `compose`? Tutaj tylko gdybam - nie wiem czy to jest możliwe :)

Jak radzicie sobie w podobnych przypadkach?

**Comment by ptrstpp950** [comment]
Jak najbardziej możesz użyć Docker Compose. Przykład https://gopensource.com/azure-functions-express-running-azure-functions-locally-using-docker-compose-bf6be03250fc
Acz ja często miałem Powershell/Bash który robił te kilka func start w osobnych okienkach i nie było to upierdliwe aż tak. Możesz też odpalić je jako background process. Wtedy logi się mieszają i też jest śmiesznie ;)

**Reply by mmisztal1980** [reply]
Możesz użyć zarówno docker-compose jak i tye z `tye run --watch` 😉

**Reply by GKotfis** [reply]
Muszę przyznać, że z `Tye` poszło szybko :) Nawet jest jakiś panel www dla mniej ogarniętych.
Jedyne co mnie zastanawia to, czy da radę jakoś pożenić `local.settings.json` tak, żeby port i  zmienne środowiskowe brał stamtąd.
Żeby uniknąć duplikowania i deklaracji tego w `tye.yaml`
Chyba zaproponuje obydwa podejścia. Plus jedno z lawiną okienek :)

**Reply by mmisztal1980** [reply]
@GKotfis powinno się udać... w gruncie rzeczy tye tylko stawia [proxy](https://github.com/dotnet/tye/blob/main/src/Microsoft.Tye.Proxy/Program.cs) w dockerze do komunikacji z procesem poza docker'em.... spróbuj wczytać te `local.settings.json`

**Reply by GKotfis** [reply]
Jednak to Tye jest w zbyt wczesnej wersji. Nie mogę uzyskać wsparcia dla podstawienia zmiennych środowiskowych.
Próba zadeklarowania `env` czy `env_file` kończy się błędem:

```
Drats! 'run' failed:
        Unable to determine service type.
```

Tak wygląda manifest:
```yaml
name: az-functions
services:
- name: api-hiring
  azureFunction: api/MyFunction
  env_file:
    - env_shared.env
````
